### PR TITLE
ci: extract addFinalDockerImage into shell script

### PR DESCRIPTION
**Pull request by keegancsmith**
This makes it much clearer what the step is doing. We also include some
buildkite log separators for easier reading of the output.

### CODE REVIEW

File: enterprise/dev/ci/ci/pipeline-steps.go [hunk 1]
```diff
@@ -348,45 +348,28 @@ func addCandidateDockerImage(c Config, app string) func(*bk.Pipeline) {
...
-			bk.Cmd("yes | gcloud auth configure-docker"),
-		}
-
 		gcrImage := fmt.Sprintf("us.gcr.io/sourcegraph-dev/%s", strings.TrimPrefix(baseImage, "sourcegraph/"))
-
-		candidateImage := fmt.Sprintf("%s:%s", gcrImage, candidateImageTag(c))
-		cmds = append(cmds,
-			bk.Cmd(fmt.Sprintf("docker pull %s", candidateImage)),
-			bk.Cmd(fmt.Sprintf("docker tag %s %s:%s", candidateImage, baseImage, c.version)),
-		)
```
Comment (keegancsmith) [inline enterprise/dev/ci/ci/pipeline-steps.go:363]: This step is removed. We don't need to locally tag the image first. Instead we can just tag the candidate image we pulled for each image we want to push.

**Comment by keegancsmith** [comment]
I just added another commit which includes a little summary. I often look at buildkite builds to try and find the images that are pushed. This now includes a simple summary in the output (with sha256) of the images published :)

Here is some example output (but I commented out the part in the script where we push)

```shellsession
$ ./docker-publish.sh \
  us.gcr.io/sourcegraph-dev/searcher:358c739f2c8ba8618c4da31dfa08c08b38777164_71747_candidate \
  index.docker.io/sourcegraph/searcher:71747_2020-08-25_358c739 \
  index.docker.io/sourcegraph/searcher:insiders \
  us.gcr.io/sourcegraph-dev/searcher:71747_2020-08-25_358c739 \
  us.gcr.io/sourcegraph-dev/searcher:insiders

--- pulling us.gcr.io/sourcegraph-dev/searcher:358c739f2c8ba8618c4da31dfa08c08b38777164_71747_candidate
358c739f2c8ba8618c4da31dfa08c08b38777164_71747_candidate: Pulling from sourcegraph-dev/searcher
df20fa9351a1: Already exists
57a7d5ff3995: Pull complete
f779407d7d74: Pull complete
b353c65d6248: Pull complete
26dd9177a1fd: Pull complete
a8fa8ee6de8b: Pull complete
2050e792270e: Pull complete
a038637ff46d: Pull complete
Digest: sha256:da3e1322a3cab2c4570ffca39e3c0508cb938488d42945455051a59bcaaf1abe
Status: Downloaded newer image for us.gcr.io/sourcegraph-dev/searcher:358c739f2c8ba8618c4da31dfa08c08b38777164_71747_candidate
us.gcr.io/sourcegraph-dev/searcher:358c739f2c8ba8618c4da31dfa08c08b38777164_71747_candidate
--- pushing index.docker.io/sourcegraph/searcher:71747_2020-08-25_358c739
--- pushing index.docker.io/sourcegraph/searcher:insiders
--- pushing us.gcr.io/sourcegraph-dev/searcher:71747_2020-08-25_358c739
--- pushing us.gcr.io/sourcegraph-dev/searcher:insiders
--- summary
index.docker.io/sourcegraph/searcher:71747_2020-08-25_358c739@sha256:866418606e18ea464329764bcbfe21a5a9e708bb9ede77a8bd317f96d1f5c71a
index.docker.io/sourcegraph/searcher:insiders@sha256:866418606e18ea464329764bcbfe21a5a9e708bb9ede77a8bd317f96d1f5c71a
us.gcr.io/sourcegraph-dev/searcher:71747_2020-08-25_358c739@sha256:866418606e18ea464329764bcbfe21a5a9e708bb9ede77a8bd317f96d1f5c71a
us.gcr.io/sourcegraph-dev/searcher:insiders@sha256:866418606e18ea464329764bcbfe21a5a9e708bb9ede77a8bd317f96d1f5c71a
```

File: dev/ci/docker-publish.sh [hunk 2]
```diff
@@ -0,0 +1,38 @@
+#!/usr/bin/env bash
+
+# This script will fetch an image and then push it to multiple different
+# locations. This is used in our CI pipeline. We build a private candidate
+# image and once e2e test pass we publish it to dockerhub.
```
Comment (bobheadxi) [inline dev/ci/docker-publish.sh:5]: does this mean we build -> push -> pull -> tag -> push? Is the built image not available on the machine without pulling it again?

**Review by bobheadxi — APPROVED** [review]
🐳 

one none-blocking question

File: dev/ci/docker-publish.sh (hunk 2 above)
Comment (bobheadxi) [reply dev/ci/docker-publish.sh:5]: curious if https://github.com/sourcegraph/sourcegraph/pull/13320/files#diff-0bbbc78c298106eb1d9d33e7d654f677R25 is actually needed

Comment (keegancsmith) [reply dev/ci/docker-publish.sh:5]: Yeah in our CI we build and push the candidate images to a private registry. Then in a later step (this one) we pull the candidate image and then push them to our public registries with different tags. The example command in the script is an example of what we would actually run in this step in CI.

Comment (keegancsmith) [reply dev/ci/docker-publish.sh:5]: > curious if https://github.com/sourcegraph/sourcegraph/pull/13320/files#diff-0bbbc78c298106eb1d9d33e7d654f677R25 is actually needed

it is, the agent that builds the candidate image is not necessarily the same as the agent which runs this step. So won't necessarily have the image. If it does have the image, then the docker pull should mostly be a noop :)

**Comment by keegancsmith** [comment]
aaah, much better

image

File: dev/ci/docker-publish.sh (hunk 2 above)
Comment (pecigonzalo) [reply dev/ci/docker-publish.sh:5]: My 2c as well, that is the recommended workflow for CI/CD as well, there is a build and push step and all other steps should pull the image/binary and do tests, if all the tests pass, you pull and push with new tags. We have in our pipeline statefull agents, but in general CI agents should be stateless.

**Pull request merged**